porch/controllers/downstreampackages/pkg/controllers/downstreampackage/downstreampackage_controller.go

```diff
@@ -0,0 +1,366 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package downstreampackage
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"strconv"
+	"strings"
+
+	porchapi "github.com/GoogleContainerTools/kpt/porch/api/porch/v1alpha1"
+	api "github.com/GoogleContainerTools/kpt/porch/controllers/downstreampackages/api/v1alpha1"
+	"golang.org/x/mod/semver"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	configapi "github.com/GoogleContainerTools/kpt/porch/api/porchconfig/v1alpha1"
+)
+
+type Options struct{}
+
+func (o *Options) InitDefaults()                       {}
+func (o *Options) BindFlags(_ string, _ *flag.FlagSet) {}
+
+// DownstreamPackageReconciler reconciles a DownstreamPackage object
+type DownstreamPackageReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	Options
+
+	packageRevs porchapi.PackageRevisionList
+}
+
+const (
+	workspaceNamePrefix = "downstreampackage-"
+)
+
+//go:generate go run sigs.k8s.io/controller-tools/cmd/controller-gen@v0.8.0 rbac:roleName=porch-controllers-downstreampackages webhook paths="." output:rbac:artifacts:config=../../../config/rbac
+
+//+kubebuilder:rbac:groups=config.porch.kpt.dev,resources=downstreampackages,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=config.porch.kpt.dev,resources=downstreampackages/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=config.porch.kpt.dev,resources=downstreampackages/finalizers,verbs=update
+
+// Reconcile implements the main kubernetes reconciliation loop.
+func (r *DownstreamPackageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	dp, err := r.init(ctx, req)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if dp == nil {
+		// maybe the dp was deleted
+		return ctrl.Result{}, nil
+	}
+
+	upstream := r.getUpstreamPR(dp.Spec.Upstream)
+	if upstream == nil {
+		return ctrl.Result{}, fmt.Errorf("could not find upstream package revision")
+	}
+
+	if err := r.ensureDownstreamPackage(ctx, dp, upstream); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// TODO: Prune (propose deletion of) deployment packages created by this controller
+	//   that are no longer needed. Part of this will be to implement the DeletionPolicy
+	//   field, which will allow the user to specify whether to "disown" or "delete".
+
+	return ctrl.Result{}, nil
+}
+
+func (r *DownstreamPackageReconciler) init(ctx context.Context, req ctrl.Request) (*api.DownstreamPackage, error) {
+	var dp api.DownstreamPackage
+
+	if err := r.Client.Get(ctx, req.NamespacedName, &dp); err != nil {
+		return nil, client.IgnoreNotFound(err)
+	}
+
+	var prList porchapi.PackageRevisionList
+	if err := r.Client.List(ctx, &prList, client.InNamespace(dp.Namespace)); err != nil {
+		return nil, err
+	}
+	r.packageRevs = prList
```

Comment:
I suggest passing this list around instead of using a property on the reconciler. I think we only use a single reconcile routine, but we usually try to avoid having the controllers stateful.

Reply:
done